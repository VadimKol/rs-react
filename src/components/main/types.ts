import { type RefObject } from 'react';
import { type Character } from 'rickmortyapi';

export interface MainProps {}

export interface MainState {
  characters: Character[];
  total: number;
  page: number;
  loader: boolean;
  character: { name: string };
  err: string;
  searchField: RefObject<HTMLInputElement>;
}
