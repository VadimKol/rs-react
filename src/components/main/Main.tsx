import { type MouseEventHandler, type ReactNode, useEffect, useRef, useState } from 'react';
import { Outlet, useLocation, useNavigate, useSearchParams } from 'react-router-dom';
import { type Character, getCharacters } from 'rickmortyapi';

import { useLocalStorage } from '@/hooks/useLocalStorage';

import { NoMatch } from '../no-match/NoMatch';
import { Results } from '../results/Results';
import { Search } from '../search/Search';
import styles from './styles.module.scss';

export function Main(): ReactNode {
  const [pageQuery, setPageQuery] = useSearchParams();
  const [characters, setCharacters] = useState<Character[]>([]);
  const [total, setTotal] = useState(0);
  const [page, setPage] = useState(Math.floor(Number(pageQuery.get('page'))) || 1);
  const [loader, setLoader] = useState(true);
  const [ls] = useLocalStorage('R&M_search');
  const [character, setCharacter] = useState({ name: ls });
  const [err, setErr] = useState('');
  const [noMatch, setNoMatch] = useState(false);
  const searchField = useRef<HTMLInputElement>(null);
  const { pathname } = useLocation();
  const characterID = pathname.replace('/character/', '');
  const navigate = useNavigate();

  if (!pageQuery.get('page')) {
    setPageQuery({ page: String(page) }, { replace: true });
  }

  if (err) {
    throw new Error(err);
  }

  if (characterID !== '/' && (/\D/.test(characterID) || characterID.startsWith('0'))) {
    setNoMatch(true);
  }

  useEffect(() => {
    if (!noMatch) {
      getCharacters({ page, name: character.name })
        .then((response) => {
          if (response.status === 200 && response.data.results) {
            setCharacters(response.data.results);
            setTotal(response.data.info?.pages || 0);
          } else if (response.status === 404) {
            setCharacters([]);
            setTotal(0);
          } else {
            throw new Error(response.statusMessage);
          }
        })
        .catch((error: Error) => setErr(error.message))
        .finally(() => setLoader(false));
    }
  }, [page, character, noMatch]);

  if (noMatch) {
    return <NoMatch setNoMatch={setNoMatch} />;
  }

  const handleClose: MouseEventHandler = (e) => {
    if (e.target === e.currentTarget && characterID !== '/') {
      navigate('/');
    }
  };

  return (
    <main className="main" onClick={handleClose} role="presentation">
      <section className={styles.search} onClick={handleClose}>
        <Search
          character={character}
          loader={loader}
          searchField={searchField}
          setCharacter={setCharacter}
          setPage={setPage}
          setLoader={setLoader}
        />
        <button className={styles.error} type="button" onClick={() => setErr('Oops, something went wrong!')}>
          Error
        </button>
      </section>
      <section className={styles.results} onClick={handleClose}>
        <div className={characterID !== '/' ? styles.character : styles.results_box}>
          {loader ? (
            <div className={styles.loader} />
          ) : (
            <Results
              characters={characters}
              total={total}
              page={page}
              setPage={setPage}
              setLoader={setLoader}
              characterID={characterID}
              handleClose={handleClose}
            />
          )}
          <Outlet context={{ characterID, setNoMatch }} />
        </div>
      </section>
    </main>
  );
}
