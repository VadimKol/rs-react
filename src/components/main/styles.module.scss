.search,
.results {
  display: flex;
  justify-content: center;
  width: 100%;
}

.search {
  gap: 200px;
  align-items: center;
  margin-bottom: 20px;

  @media screen and (width <= 768px) {
    flex-direction: column-reverse;
    gap: 20px;
  }
}

.error {
  width: 80px;
  height: 80px;
  font-size: 30px;
  background-color: var(--error-color);
  border: none;
  border-radius: 100%;
  transition: box-shadow 0.3s, transform 0.3s;

  &:focus-visible {
    outline: 2px solid var(--primary-color);
  }

  &:active {
    transform: scale(0.9);
  }

  @media (hover: hover) {
    &:hover {
      box-shadow: 0 0 5px 5px rgb(196 35 35 / 51.4%);
      cursor: pointer;
    }
  }
}

.results {
  flex-grow: 1;
  flex-direction: column;
  overflow: hidden;
}

.loader {
  flex-grow: 1;
  width: 100%;
  background-image: url("/src/assets/images/loader.png");
  background-position: center;
  background-size: contain;
  background-repeat: no-repeat;
  animation: loader 1s infinite linear;
}

@keyframes loader {
  to {
    transform: rotate(1turn);
  }
}
