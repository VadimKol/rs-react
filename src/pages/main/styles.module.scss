.links {
  display: flex;
  gap: 1.25rem;
}

.link {
  position: relative;
  padding: 0.5em 1.2em;
  text-decoration: none;
  color: var(--text-color);
  background:
    linear-gradient(45deg, var(--primary-color), var(--secondary-color)) padding-box,
    linear-gradient(45deg, var(--primary-color), var(--secondary-color)) border-box;
  border: none;
  border-radius: 6px;
  user-select: none;
  transition: transform 0.3s, box-shadow 0.3s;
  transform-style: preserve-3d;

  &::before {
    content: "";
    position: absolute;
    inset: 0;
    background: linear-gradient(45deg, var(--primary-color), var(--secondary-color));
    border-radius: inherit;
    filter: blur(0);
    pointer-events: none;
    transition: all 0.5s;
    transform: translate3d(0, 0, -1px);
  }


  &:focus-visible {
    outline: 2px solid var(--primary-color);
  }

  &:active {
    transform: scale(0.9);
  }

  @media (hover: hover) {
    &:hover {
      cursor: pointer;

      &::before {
        filter: blur(5px);
      }
    }
  }
}

.links_loader {
  margin-bottom: 1rem;
}

.uncontrolled {
  position: relative;
  display: flex;
  flex-grow: 1;
  flex-direction: column;
  gap: 0.63rem;
  width: 37.5rem;
  height: 100%;
  margin-top: 1rem;
  padding: 1.25rem;
  color: var(--background-color);
  background:
    linear-gradient(var(--text-color) 0 0) padding-box,
    linear-gradient(to top, var(--primary-color), var(--secondary-color)) border-box;
  border: 5px solid transparent;
  border-radius: 10px;
  transform-style: preserve-3d;

  &::before {
    content: "";
    position: absolute;
    inset: 0;
    background: linear-gradient(to top, var(--primary-color), var(--secondary-color));
    border-radius: inherit;
    filter: blur(10px);
    pointer-events: none;
    transition: all 0.3s;
    transform: translate3d(0, 0, -1px);
  }
}

.title {
  text-align: center;
}

.form_info {
  display: flex;
  flex-direction: column;
  gap: 1rem;
}

.text {
  margin: 0;
}
