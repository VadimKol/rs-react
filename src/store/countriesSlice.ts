import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';

import { countries } from '@/common/countries';

import { type RootState } from './store';

export const countriesLice = createSlice({
  name: 'countries',
  initialState: countries,
  reducers: {
    addCountry: (state, action: PayloadAction<string>) => {
      state.push(action.payload);
    },
  },
});

export const { addCountry } = countriesLice.actions;

export const selectCountries = (state: RootState): string[] => state.countries;

export const useCountries = (): string[] => useSelector(selectCountries);
