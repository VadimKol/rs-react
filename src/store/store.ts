import { configureStore } from '@reduxjs/toolkit';

import { formSLice } from './formSLice';

export const store = configureStore({ reducer: { form: formSLice.reducer } });

export type RootState = ReturnType<typeof store.getState>;
