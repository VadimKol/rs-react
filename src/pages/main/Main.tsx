import { nanoid } from '@reduxjs/toolkit';
import { type ReactNode, useEffect, useState } from 'react';
import { Link, useLocation } from 'react-router-dom';

import { getPasswordStrength } from '@/common/utils';
import { ImageBlock } from '@/components/image-block/ImageBlock';
import { Loader } from '@/components/loader/Loader';
import { useFormInfo } from '@/store/formSLice';

import styles from './styles.module.scss';

export function Main(): ReactNode {
  const theme = 'dark';
  const forms = useFormInfo();
  const [loader, setLoader] = useState(false);
  const location = useLocation();

  useEffect(() => {
    if (location.state) {
      setLoader(true);
    }
  }, [location.state]);

  useEffect(() => {
    if (loader) {
      setTimeout(() => {
        setLoader(false);
      }, 2000);
    }
  }, [loader]);

  return (
    <main className={theme === 'dark' ? 'main' : 'main light'}>
      <section className={loader ? `${styles.links} ${styles.links_loader}` : styles.links}>
        <Link className={styles.link} to="/uncontrolled">
          Uncontrolled Form
        </Link>
        <Link className={styles.link} to="/rhf">
          React hook form
        </Link>
      </section>
      {forms.map((form, index) =>
        index === forms.length - 1 && loader ? (
          <Loader key={`${nanoid()}`} />
        ) : (
          <div key={`${nanoid()}`} className={loader ? `${styles.forms} ${styles.forms_loader}` : styles.forms}>
            <section className={styles.uncontrolled}>
              <div className={styles.form_info}>
                <p className={styles.text}>Name: {form.name}</p>
                <p className={styles.text}>Age: {form.age}</p>
                <p className={styles.text}>Email: {form.email}</p>
                <p className={styles.text}>Password: {'*'.repeat(form.password?.length || 0)}</p>
                <p className={styles.text}>Password strength: {getPasswordStrength(form.strength)}</p>
                <p className={styles.text}>Gender: {form.gender}</p>
                <p className={styles.text}>Country: {form.country}</p>
                <p className={styles.text}>TC: agreed</p>
                <ImageBlock src={form.image || ''} alt="Form data" />
              </div>
            </section>
          </div>
        ),
      )}
    </main>
  );
}
