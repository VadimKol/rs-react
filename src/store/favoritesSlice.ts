/* import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';
import { type Character } from 'rickmortyapi';

import { type RootState } from './store';

const initialState: Character[] = [];

export const favoritesSLice = createSlice({
  name: 'favorites',
  initialState,
  reducers: {
    addToFavorites: (state, action: PayloadAction<Character>) => {
      state.push(action.payload);
    },
    removeFromFavorites: (state, action: PayloadAction<Character>) =>
      state.filter((character) => character.id !== action.payload.id),
    clearFavorites: () => [],
  },
});

export const { addToFavorites, removeFromFavorites, clearFavorites } = favoritesSLice.actions;

const selectFavorites = (state: RootState): Character[] => state.favorites;

const isFavorite = (state: RootState, character?: Character): boolean =>
  state.favorites.some((favorite) => favorite.id === character?.id);

export const useFavorites = (): Character[] => useSelector(selectFavorites);

export const useIsFavorite = (character?: Character): boolean =>
  useSelector((state: RootState) => isFavorite(state, character));
 */
