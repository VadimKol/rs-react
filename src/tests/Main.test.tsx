import { createRemixStub } from '@remix-run/testing';
import { render } from '@testing-library/react';
import { type ReactNode } from 'react';
import { Provider } from 'react-redux';

import { Main } from '@/components/main/Main';
import { store } from '@/store/store';

import { characters } from './mocks/data';

global.URL.createObjectURL = vi.fn();
global.URL.revokeObjectURL = vi.fn();

describe('Main page', () => {
  it('renders correctly', () => {
    const RemixStub = createRemixStub([
      {
        path: '/',
        Component: (): ReactNode => (
          <Provider store={store}>
            <Main charactersData={{ characters, total: 1 }} />
          </Provider>
        ),
      },
    ]);

    const { container } = render(<RemixStub />);

    expect(container).toMatchSnapshot();
  });
});
