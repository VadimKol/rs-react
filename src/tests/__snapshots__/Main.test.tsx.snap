// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Main page > renders correctly 1`] = `
<div>
  <main
    class="main"
    role="presentation"
  >
    <section>
      <form
        class="_search_form_5ff5ab"
      >
        <input
          class="_search_5ff5ab"
          id="search"
          placeholder="Search..."
          type="text"
          value=""
        />
        <button
          aria-label="Search-button"
          class="_search_button_5ff5ab"
          type="submit"
        />
      </form>
    </section>
    <section
      class="_results_24fb31"
    >
      <div
        class="_results_box_24fb31"
      >
        <div
          class="_no_character_04010f"
        >
          <ul
            class="_characters_04010f"
            role="presentation"
          >
            <li
              class="_characters_item_04010f"
            >
              <div
                class="_card_0086a2"
              >
                <div
                  class="_image_container_de2c00"
                >
                  <div
                    class="_loader_d8e6ec"
                  />
                  <img
                    alt="Character"
                    class="_hidden_de2c00"
                    height="300"
                    src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
                    width="300"
                  />
                </div>
                <div
                  class="_text_container_0086a2"
                >
                  <a
                    class="_card_link_0086a2"
                    data-discover="true"
                    href="/character/1?page=1&search="
                  >
                    <h2
                      class="_card_title_0086a2"
                    >
                      Rick
                    </h2>
                  </a>
                  <button
                    aria-label="Add to favorites"
                    class="_favorite_99b3bf"
                    name="favorite"
                    type="button"
                    value="false"
                  >
                    ☆
                  </button>
                </div>
              </div>
            </li>
            <li
              class="_characters_item_04010f"
            >
              <div
                class="_card_0086a2"
              >
                <div
                  class="_image_container_de2c00"
                >
                  <div
                    class="_loader_d8e6ec"
                  />
                  <img
                    alt="Character"
                    class="_hidden_de2c00"
                    height="300"
                    src="https://rickandmortyapi.com/api/character/avatar/2.jpeg"
                    width="300"
                  />
                </div>
                <div
                  class="_text_container_0086a2"
                >
                  <a
                    class="_card_link_0086a2"
                    data-discover="true"
                    href="/character/2?page=1&search="
                  >
                    <h2
                      class="_card_title_0086a2"
                    >
                      Morty
                    </h2>
                  </a>
                  <button
                    aria-label="Add to favorites"
                    class="_favorite_99b3bf"
                    name="favorite"
                    type="button"
                    value="false"
                  >
                    ☆
                  </button>
                </div>
              </div>
            </li>
          </ul>
          <div
            class="_pagination_34bd36"
          >
            <div
              class="_pagination_box_34bd36"
            >
              <div
                class="_pages_34bd36"
              >
                <span>
                  1
                </span>
                -
                <span>
                  1
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </section>
  </main>
</div>
`;
